datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -86,45 +86,54 @@ func NewChainDataFetcher(ctx *node.ServiceContext, cfg *ChainDataFetcherConfig)
 	var (
 		repo         Repository
 		checkpointDB CheckpointDB
-		setter       ComponentSetter
+		setters      []ComponentSetter
 		err          error
 	)
 	switch cfg.Mode {
 	case ModeKAS:
-		newRepo, err := kas.NewRepository(cfg.KasConfig)
+		repo, checkpointDB, setters, err = getKasComponents(cfg.KasConfig)
 		if err != nil {
-			logger.Error("Failed to create new Repository", "err", err, "user", cfg.KasConfig.DBUser, "host", cfg.KasConfig.DBHost, "port", cfg.KasConfig.DBPort, "name", cfg.KasConfig.DBName, "cacheUrl", cfg.KasConfig.CacheInvalidationURL, "x-chain-id", cfg.KasConfig.XChainId)
 			return nil, err
 		}
-		repo = newRepo
-		checkpointDB = newRepo
-		setter = newRepo
 	case ModeKafka:
-		// TODO-ChainDataFetcher implement new repository for kafka
-		panic("implement me")
+		repo, checkpointDB, setters, err = getKafkaComponents(cfg.KafkaConfig)
+		if err != nil {
+			return nil, err
+		}
 	default:
 		logger.Error("the chaindatafetcher mode is not supported", "mode", cfg.Mode)
 		return nil, errUnsupportedMode
 	}
-	checkpoint, err := checkpointDB.ReadCheckpoint()
-	if err != nil {
-		logger.Error("Failed to get checkpoint", "err", err)
-		return nil, err
-	}
 	return &ChainDataFetcher{
 		config:        cfg,
 		chainCh:       make(chan blockchain.ChainEvent, cfg.BlockChannelSize),
 		reqCh:         make(chan *cfTypes.Request, cfg.JobChannelSize),
 		stopCh:        make(chan struct{}),
 		numHandlers:   cfg.NumHandlers,
-		checkpoint:    checkpoint,
 		checkpointMap: make(map[int64]struct{}),
 		repo:          repo,
 		checkpointDB:  checkpointDB,
-		setter:        setter,
+		setters:       setters,
 	}, nil
 }
 
+func getKasComponents(cfg *kas.KASConfig) (Repository, CheckpointDB, []ComponentSetter, error) {
```

Comment:
Returning 4 variables looks verbose, please consider refactoring them later.

Reply:
Aright. I will make another PR for this.